### DISCUSSION TITLE
Incremental builds?

### DISCUSSION THREAD

**Discussion by tve** [Moddable-OpenSource/moddable#983] [General]
Is is possible to make incremental builds with the moddable sdk? E.g. where a base system is loaded and only small portions of the app are rebuilt and reloaded? Obviously this requires some form of incremental linking and loading...

I really love interpreted languages on embedded systems because of the built-in REPL capability. I don't actually use a true REPL much, if ever, but the REPL allows loading some small snippets of code for rapid experimentation as well as for troubleshooting. I know that XS uses server-side compilation and ships byte code, but that doesn't preclude a REPL-like environment on the desktop where I can type (or hit save/run in my editor) and the code gets compiled locally and then shipped to the MCU to be loaded and executed.

I'm also quite fond of having remote MQTTS-only devices (esp32, specifically) whose only comm mechanism is to an MQTTS server. (I use this almost exclusively with Micropython.) This implies quite a hefty base system, since it has to support MQTTS as well as loading and managing new code. Being able to pre-load such a base system including many of the larger stable portions of a project and then just sending small application code overlays is incredibly powerful in my experience. I actually only develop apps using this model, even when the device is sitting on my desktop...

Maybe all this already exists and I just haven't discovered it yet... Thoughts?

**Comment by phoddie** [comment]
Check out [mods](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/xs/mods.md).  They let you install precompiled JavaScript and resources on top of a base firmware image. They are used for all the examples in our [book](https://www.moddable.com/book), for example, to make it quick to try out examples.

**Reply by tve** [reply]
I don't deal with tech books very well... (Nothing about price) Thanks for your patience!

**Comment by tve** [comment]
Thanks for the pointer to the docs! I have some follow-up questions:

- In "building the mod" the doc says "The first time a mod is run by a mod host, XS automatically updates the symbol ID values in the mod's byte code so they match the host IDs." Where does "XS" run? Is this the run-time engine that performs the symbol update or the host-based tooling? I have the feeling it's the latter...
- How is the mod loading done? E.g. what is the protocol? I suspect is the xsbug protocol? Can you point me to that?
- Is there a hard reason one can't load multiple mods?

**Reply by phoddie** [reply]
> Where does "XS" run? Is this the run-time engine that performs the symbol update or the host-based tooling? I have the feeling it's the latter...

It is the former.

> How is the mod loading done? E.g. what is the protocol? I suspect is the xsbug protocol? Can you point me to that?

By "loading" I think you mean "installing"? That can be done any way you like: it is just data being transferred to a flash partition. `mcrun` sends the data under the xsbug protocol.

> s there a hard reason one can't load multiple mods?

It is supported, though not yet publicly documented. The `mcrun` installation only supports a single mod, however.

**Comment by tve** [comment]
Is there a generic mod host that runs on esp32 and gets mods over the network? What I'm looking for is a bare bones example to get started with having an esp32 remote doing nothing other than "being on the network" and being able to ship it mods to run.

**Reply by phoddie** [reply]
> Is there a generic mod host that runs on esp32 and gets mods over the network? 

No.

Fortunately, it is straightforward to do. There are commercial products out in the world that do it.
